src/expr/src/linear.rs

```diff
@@ -1761,94 +1744,65 @@ pub mod plan {
             for u in self.upper_bounds.iter() {
                 // We can cease as soon as the lower and upper bounds match,
                 // as the update will certainly not be produced in that case.
-                if upper_bound != lower_bound {
+                if upper_bound != Some(lower_bound) {
                     match u.eval(datums, &arena) {
+                        Err(EvalError::MzTimestampStepOverflow) => {
+                            // Since this upper bound exceeds all timestamps, it
+                            // does not modify `upper_bound`.
+                        }
                         Err(e) => {
                             return Some(Err((e.into(), time, diff)))
                                 .into_iter()
                                 .chain(None.into_iter());
                         }
-                        Ok(Datum::Numeric(d)) => {
-                            // If the upper bound is negative, it is impossible to satisfy.
-                            // We set the upper bound to the lower bound to indicate this.
-                            if d.0.is_negative() {
-                                upper_bound = lower_bound;
+                        Ok(Datum::MzTimestamp(d)) => {
+                            if let Some(upper) = upper_bound {
+                                upper_bound = Some(upper.min(d));
                             } else {
-                                // An `Ok` conversion is a valid upper bound, and an `Err` error
-                                // indicates a value above `u64::MAX`. The `ok()` method does the
-                                // conversion we want to an `Option<u64>`.
-                                let v = u64::try_from(d.0).ok().map(mz_repr::Timestamp::from);
-                                // Update `upper_bound` to be the minimum with `v`, where a `None`
-                                // value is treated as larger than `Some(_)` values.
-                                upper_bound = match (upper_bound, v) {
-                                    (None, x) => x,
-                                    (x, None) => x,
-                                    (x, y) => x.min(y),
-                                };
-                                // Force the upper bound to be at least the lower bound.
-                                // The `is_some()` test is required as `None < Some(_)`,
-                                // and we do not want to set `None` to `Some(_)`.
-                                if upper_bound.is_some() && upper_bound < lower_bound {
-                                    upper_bound = lower_bound;
-                                }
+                                upper_bound = Some(d);
+                            };
+                            // Force the upper bound to be at least the lower bound.
+                            if upper_bound < Some(lower_bound) {
+                                upper_bound = Some(lower_bound);
```

Comment:
Is the argument that `upper_bound` cannot be `None` here, because code paths above make it so? I'd love to put some text here to that extent, if it is correct. Someone could go and mangle the code up above and for some reason put in a `None` (idk why) and we would need to verify that `is_some()` is still true. You could leave the test in, for example, and Rust could/should be able to optimize it out.